auto_round_extension/torch/qlinear_torch.py

```diff
@@ -0,0 +1,278 @@
+                shifts = torch.arange(0, 10) * self.bits + 1  
+                shifted = packed_weight << shifts 
+                qweight[row] |= shifted.sum(dim=-1)
+                i += 10
+                qweight[row] |= intweight[i] << 31
+                row += 1
+                qweight[row] |= (intweight[i] >> 1) & 0x3
+                i += 1
+                packed_weight = torch.tensor(intweight[i : i + 10]).to(dtype=torch.int32).t()
+                shifts = torch.arange(0, 10) * self.bits + 2
+                shifted = packed_weight << shifts 
+                qweight[row] |= shifted.sum(dim=-1)
+                i += 10
+                row += 1
+                
+            self.qweight = qweight.cpu()
+            
+        if isinstance(zeros, torch.Tensor):
+            zeros = zeros.t().contiguous()
+            zeros = zeros.numpy().astype(np.uint32)
+            qzeros = torch.zeros((zeros.shape[0], zeros.shape[1] // 32 * self.bits), dtype=torch.int32)
+            i = 0
+            col = 0
+            while col < qzeros.shape[1]:
+                if self.bits in [2, 4, 8]:
+                    packed_zeros = torch.tensor(zeros[:, i: i + (32 // self.bits)]).to(dtype=torch.int32)
+                    shifts = torch.arange(0, (32 // self.bits)) * self.bits
+                    shifted = packed_zeros << shifts
+                    qzeros[:, col] |= shifted.sum(dim=-1)
+                    i += 32 // self.bits
+                    col += 1
+                elif self.bits == 3:
+                    packed_zeros = torch.tensor(zeros[:, i : i + 10]).to(dtype=torch.int32)
+                    shifts = torch.arange(0, 10) * self.bits
+                    shifted = packed_zeros << shifts                  
+                    qzeros[:, col] = shifted.sum(dim=-1)
+                    i += 10
+                    qzeros[:, col] |= zeros[:, i] << 30
+                    col += 1
+                    qzeros[:, col] |= (zeros[:, i] >> 2) & 1
+                    i += 1
+                    packed_zeros = torch.tensor(zeros[:, i : i + 10]).to(dtype=torch.int32)
+                    shifts = torch.arange(0, 10) * self.bits + 1  
+                    shifted = packed_zeros << shifts 
+                    qzeros[:, col] |= shifted.sum(dim=-1)
+                    i += 10
+                    qzeros[:, col] |= zeros[:, i] << 31
+                    col += 1
+                    qzeros[:, col] |= (zeros[:, i] >> 1) & 0x3
+                    i += 1
+                    packed_zeros = torch.tensor(zeros[:, i : i + 10]).to(dtype=torch.int32) 
+                    shifts = torch.arange(0, 10) * self.bits + 2
+                    shifted = packed_zeros << shifts
+                    qzeros[:, col] |= shifted.sum(dim=-1)
+                    i += 10
+                    col += 1
+
+            self.qzeros = qzeros.cpu()
+        else:
+            assert self.bits in [2, 4, 8]
```

Comment:
t's still unclear why this assertion is added here,  the else branch should correspond to the case where zp is not a tensor, which implies symmetric quantization.